crates/parser/src/ast.rs

```diff
@@ -386,6 +389,327 @@ impl Spanned for ContractStmt {
+                writeln!(indented(f), "{}", node_line_joined(body))
+            }
+            FuncStmt::If {
+                test,
+                body,
+                or_else,
+            } => {
+                writeln!(f, "if {}:", test.kind)?;
+                writeln!(indented(f), "{}", node_line_joined(body))?;
+                if !or_else.is_empty() {
+                    writeln!(f, "else {}:", test.kind)?;
+                    writeln!(indented(f), "{}", node_line_joined(or_else))
+                } else {
+                    Ok(())
+                }
+            }
+            FuncStmt::Assert { test, msg } => {
+                if let Some(msg) = msg {
+                    write!(f, "assert {}, {}", test.kind, msg.kind)
+                } else {
+                    write!(f, "assert {}", test.kind)
+                }
+            }
+            FuncStmt::Emit { name, args } => {
+                write!(f, "emit {}({})", name.kind, node_comma_joined(&args.kind))
+            }
+            FuncStmt::Expr { value } => write!(f, "{}", value.kind),
+            FuncStmt::Pass => write!(f, "pass"),
+            FuncStmt::Break => write!(f, "break"),
+            FuncStmt::Continue => write!(f, "continue"),
+            FuncStmt::Revert { .. } => write!(f, "revert"),
+        }
+    }
+}
+
+impl fmt::Display for VarDeclTarget {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        match self {
+            VarDeclTarget::Name(name) => write!(f, "{}", name),
+            VarDeclTarget::Tuple(elts) => write!(f, "{}", node_comma_joined(elts))
+        }
+    }
+}
+
+impl fmt::Display for Expr {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        match self {
+            Expr::Ternary {
+                if_expr,
+                test,
+                else_expr,
+            } => write!(
+                f,
+                "{} if {} else {}",
+                if_expr.kind, test.kind, else_expr.kind
+            ),
+            Expr::BoolOperation { left, op, right } => {
+                let left = maybe_fmt_inner_with_parens(self, &left.kind);
+                let right = maybe_fmt_inner_with_parens(self, &right.kind);
+                write!(f, "{} {} {}", left, op.kind, right)
+            }
+            Expr::BinOperation { left, op, right } => {
+                let left = maybe_fmt_inner_with_parens(self, &left.kind);
+                let right = maybe_fmt_inner_with_parens(self, &right.kind);
+                write!(f, "{} {} {}", left, op.kind, right)
+            }
+            Expr::UnaryOperation { op, operand } => {
+                let operand = maybe_fmt_inner_with_parens(self, &operand.kind);
+                if op.kind == UnaryOperator::Not {
+                    write!(f, "{} {}", op.kind, operand)
+                } else {
+                    write!(f, "{}{}", op.kind, operand)
+                }
+            },
+            Expr::CompOperation { left, op, right } => {
```

Comment:
do these need parens?

Reply:
I dont think so

Reply:
Yeah, eg `(1 == 2) == (3 == 4)`

Reply:
`(a == b) == (c == d)` is equivalent to  `1 == 2 == (3 == 4)`, right? this is at least what the test are telling me, as the first set of parens are dropped when printing